src/player.ts

```diff
@@ -186,14 +186,35 @@ export default class Player {
     this.data.power.gaia = 0;
   }
 
-  buildingValue( building: Building, planet: Planet ){ 
+  buildingValue(building: Building, planet: Planet){ 
     const addedBescods = this.faction === Faction.Bescods && this.data[Building.PlanetaryInstitute] === 1  && planet === Planet.Titanium ? 1 : 0;
     //TODO value if TECH3
     return stdBuildingValue(building);
   }
 
-  maxLeech( possibleLeech: number ){ 
+  maxLeech(possibleLeech: number){ 
     // considers real chargeable power and victory points
     return Math.min( this.data.chargePower(possibleLeech, true), this.data.victoryPoints + 1);
   }
+
+  possibleResearchAreas( cost: string, destResearchArea?: ResearchField){
+    const tracks = [];
+
+    if (this.data.canPay(Reward.parse(cost))) {
+      for (const field of Object.values(ResearchField)) {
+        if ( ((destResearchArea && destResearchArea === field) || !destResearchArea) &&
```

Comment:
The `if` are a little big.

`(destResearchArea && destResearchArea === field) || !destResearchArea)` can be simplified to `(!destResearchArea || destResearchArea === field)`.

I don't think `|| this.data.research[field] === 5` is correct, as you can't upgrade when you're already at the end of the track.

If that part is removed, then you can also remove `this.data.research[field] === 4`.

There are also a `lastTile` (can be renamed) and a `keyNeeded` functions in `research-tracks.ts`. Then you can avoid testing numbers like `4` or `5`.

You can choose to simplify the ifs, or even separate them like this (personal preference):

```ts
const destTile = this.data.research[field] + 1;

// end of the track reached
if (lastTile(destResearchArea) < destTile) {
  continue;
}

if (destResearchArea && destResearchArea !== field) {
  continue;
}

// To go from 4 to 5, we need to flip a federation
if (keyNeeded(destTile) && this.data.greenFederations === 0) {
  continue;
}

tracks.push(...);
```

Reply:
Don't worry about not helping in this period, you already did a lot previously! 
As far as the tech level 5. You can actually upgrade to level 5 even if you already are in level 5. This happens if you take the corresponding tile in the research area at the very end.
I will refactored the code with your suggestions. I also have to check if some other player already entered in it.